auth.go

```diff
@@ -512,12 +565,34 @@ func (x *Central) authenticate(identity, password string) (AuthUser, error) {
 		if err == ErrInvalidCredentials {
 			// The user already exists on our system, which means it exists on LDAP due to our Merge, with
 			// that knowledge we can say the password is invalid
-			return user, ErrInvalidPassword
+			authErr = ErrInvalidPassword
 		}
-		return user, err
+		authErr = err
 	} else {
-		return user, x.userStore.Authenticate(identity, password, authTypeCheck)
+		authErr = x.userStore.Authenticate(identity, password, authTypeCheck)
 	}
+
+	if authErr == ErrInvalidPassword {
+		username := user.Username
+		x.Stats.IncrementInvalidPasswords(x.Log)
+		x.Stats.IncrementInvalidPasswordHistory(x.Log, username, clientIPAddress)
+		if x.EnableAccountLocking {
+			if isLockable, lockabilityErr := x.LockingPolicy.IsLockable(identity); lockabilityErr == nil {
+				if int(x.Stats.UserLoginAttempts[username]) >= x.MaxFailedLoginAttempts && isLockable {
```

Comment:
Need to be holding the mutex

Reply:
Rather do the mutex lock/unlock as above, where you hold it for as short as possible. In this formulation, you're holding the mutex while you do LockAccount, and AuditUserAction. Both of those calls could potentially take a LONG TIME. During that time, nobody else can login. So our "feature" has turned into a "bug".